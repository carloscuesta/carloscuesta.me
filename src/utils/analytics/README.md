# Analytics

This file defines the analytics functions used at the website
