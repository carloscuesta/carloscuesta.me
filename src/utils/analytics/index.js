// @flow
export const GA_TRACKING_ID: string = 'UA-67824860-3'

export const trackPageView = (url: string) => {
  window.ga('set', 'page', url)
  window.ga('send', 'pageview')
}
