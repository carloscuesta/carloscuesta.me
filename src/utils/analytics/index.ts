import getTrackingCode from './tracking'

const GA_TRACKING_ID = 'UA-67824860-3'

export const trackingCode = getTrackingCode(GA_TRACKING_ID)

type AnalyticsEvent = {
  action: string,
  category: string,
  label: string,
  nonInteraction?: boolean,
  value: number
}

declare global {
  interface Window {
    ma?: {
      trackEvent?: (
        category: AnalyticsEvent['category'],
        action: AnalyticsEvent['action'],
        label: AnalyticsEvent['label'],
        value: AnalyticsEvent['value'],
        nonInteraction?: AnalyticsEvent['nonInteraction'],
      ) => void 
    };
  }
}

export const trackEvent = (event: AnalyticsEvent) => {
  if (typeof window?.ma?.trackEvent === 'function') {
    window.ma.trackEvent(
      event.category,
      event.action,
      event.label,
      event.value,
      event.nonInteraction
    )
  }
}
