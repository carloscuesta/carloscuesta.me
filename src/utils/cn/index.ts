import { cx } from 'classix'
import { twMerge } from 'tailwind-merge'

const cn = (...args: Array<string | boolean | undefined>) =>
  twMerge(cx(...args))

export default cn
