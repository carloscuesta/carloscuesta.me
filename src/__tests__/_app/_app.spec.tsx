import renderer from 'react-test-renderer'

import App, { reportWebVitals } from 'src/pages/_app'
import { trackEvent } from 'src/utils/analytics'
import * as stubs from './stubs'

jest.mock('next-seo', () => ({ DefaultSeo: 'DefaultSeo' }))

jest.mock('src/utils/analytics')

/*
  I'm mocking the components since those are unit tested.
  So there's no point to test the rendering of them twice.
  It's more easier to manage little snapshots than big ones.
*/

jest.mock('src/components/shared/Layout', () => 'Layout')

describe('_app', () => {
  describe('page', () => {
    it('should match App page', () => {
      const wrapper = renderer.create(
        // @ts-expect-error We don't need to pass router to test the App component.
        <App {...stubs.props} />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('reportWebVitals', () => {
    it('should call trackEvent with the metric as argument', () => {
      reportWebVitals(stubs.webVitalMetric)

      expect(trackEvent).toHaveBeenCalledWith({
        action: stubs.webVitalMetric.name,
        category: 'Web Vitals',
        label: stubs.webVitalMetric.id,
        value: stubs.webVitalMetric.value,
        nonInteraction: true
      })
    })
  })
})
