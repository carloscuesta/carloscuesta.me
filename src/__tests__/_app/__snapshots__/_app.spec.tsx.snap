// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_app page should match App page 1`] = `
[
  <script
    dangerouslySetInnerHTML={
      {
        "__html": "!function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&true)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}if(e==='light'||e==='dark')d.style.colorScheme=e}catch(e){}}()",
      }
    }
  />,
  <Layout>
    <DefaultSeo
      description="I'm a Front End Engineer from Barcelona. Focused on building products with JavaScript. Interested in design, under engineering and open source."
      openGraph={
        {
          "images": [
            {
              "alt": "Carlos Cuesta",
              "height": 600,
              "url": "https://carloscuesta.me/images/twitter-card.png",
              "width": 1200,
            },
          ],
          "locale": "en_GB",
          "site_name": "Carlos Cuesta",
          "type": "website",
          "url": "https://carloscuesta.me",
        }
      }
      title="Carlos Cuesta â€“ Front End Engineer"
      twitter={
        {
          "cardType": "summary_large_image",
          "handle": "@crloscuesta",
          "site": "@crloscuesta",
        }
      }
    />
    <div
      test=""
    >
      Component
    </div>
  </Layout>,
]
`;
